src/analysis/inner_product_space/orientation.lean

```diff
@@ -0,0 +1,60 @@
+/-
+Copyright (c) 2022 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Myers
+-/
+import analysis.inner_product_space.projection
+import linear_algebra.orientation
+
+/-!
+# Orientations of real inner product spaces.
+
+This file provides definitions and proves lemmas about orientations of real inner product spaces.
+
+## Main definitions
+
+* `orientation.fin_orthonormal_basis` is an orthonormal basis, indexed by `fin n`, with the given
+orientation.
+
+-/
+
+noncomputable theory
+
+variables {E : Type*} [inner_product_space ℝ E]
+variables {ι : Type*} [fintype ι] [decidable_eq ι]
+
+open finite_dimensional
+
+/-- `basis.adjust_to_orientation`, applied to an orthonormal basis, produces an orthonormal
+basis. -/
+lemma orthonormal.orthonormal_adjust_to_orientation [nonempty ι] {e : basis ι ℝ E}
+  (h : orthonormal ℝ e) (x : orientation ℝ E ι) : orthonormal ℝ (e.adjust_to_orientation x) :=
+h.orthonormal_of_forall_eq_or_eq_neg (e.adjust_to_orientation_apply_eq_or_eq_neg x)
+
+/-- An orthonormal basis, indexed by `fin n`, with the given orientation. -/
+protected def orientation.fin_orthonormal_basis {n : ℕ} (hn : 0 < n) (h : finrank ℝ E = n)
+  (x : orientation ℝ E (fin n)) : basis (fin n) ℝ E :=
+begin
+  haveI := fin.pos_iff_nonempty.1 hn,
+  haveI := finite_dimensional_of_finrank (h.symm ▸ hn : 0 < finrank ℝ E),
+  exact (fin_orthonormal_basis h).adjust_to_orientation x
+end
```

Comment:
This feels kind of awkward vs
```suggestion
protected def orientation.fin_orthonormal_basis {n : ℕ} [finite_dimensional 𝕜 E] (h : finrank ℝ E = n)
  (x : orientation ℝ E (fin n)) : basis (fin n) ℝ E :=
(fin_orthonormal_basis h).adjust_to_orientation x
```
which also has the advantage of working when `n=0`

Reply:
`adjust_to_orientation` requires a `nonempty` index type, since in zero dimensions there are two orientations but only one basis so it's not possible to produce a basis in an arbitrary orientation.

The reason for deducing `nonempty` and `finite_dimensional` type classes as part of definitions / proofs rather than including them as (formally redundant) hypotheses is that one of the reviews of one of my previous orientations PRs involved removing `nontrivial` hypotheses that could be deduced from another hypothesis and using `nontrivial_of_lt` in the proofs instead, so I took it to be the mathlib style not to include a type class hypothesis if it can be deduced from another hypothesis.

Reply:
A related question: is `orientation ℝ E (fin 0)` empty? If so, you can drop the `hn : 0 < n` hypothesis

Reply:
No, it's not empty. There are two orientations in any finite-dimensional case, including in the zero-dimensional case. In zero dimensions, one orientation is given by the determinant map with respect to the unique basis (with zero basis vectors), and the other orientation, its negation, is not given by any basis.